WIP: move job to batch/v1

erictune opened the pull request:
I am trying to move job to batch/v1.

My goals are:
1. keep supporting resource path like `/apis/extensions/v1beta1/.../jobs`
2. new resource path like `/apis/batch/v1/.../jobs`
3. both `/apis/extensions/v1/.../jobs/foo` and `/apis/batch/v1/.../jobs/foo` map to same etcd storage path, allowing easy upgrade and a single controller for both resources.
4. There will be different validation and defaulting for the new path.
5. Eventually, the old path will be deprecated.

Because of goal 3, there will only be 1 internal type for Job.
It can't live in two places.  So, I am trying to keep it in `pkg/apis/extensions` for now.

My overall plan is:
1. define new serialized representation in `pkg/apis/batch/v1` and install it.  Not exposed by default yet.
2. make changes to `pkg/apis/batch/v1` type and defaulting and validation.
3. expose it by default.  Generate clients.
4. Change controller and kubectl to use new path.

I am working on step 1 right now, and getting stuck.

I am stuck.
I have defined the new serializable type in a new directory for the new group.
I now need to generate deep copies and conversions.  I need to install the new serializeable type.
I am not sure if I should have  new `pkg/apis/batch/install/install.go` for it, or edit `pkg/apis/extensions/install/install.go` to include a second API group and additional serializable version.
Advice wanted.

erictune committed:
Copy Job external representation to new dir.

This PR is just the following:

mkdir pkg/apis/batch
mkdir pkg/apis/batch/v1
S="pkg/apis/extensions/v1beta1"
D="pkg/apis/batch/v1"
files="conversion.go register.go defaults.go defaults_test.go types.go"
for f in $files
do
  cp $S/$f $D/$f
  git add $D/$f
done

Not supposed to compile.
Subsequent PRs remove non-Job types.

erictune committed:
Remove code not related to Job.

pkg/apis/batch/v1/conversion.go
- Delete all the scheme.AddConversionFuncs except the PodSpec ones, which Job presumably needs.
- Delete the field label stuff for non-Job, keeping the stuff for Job.
- Keep the podSpec conversion funcs, delete the rest.

pkg/apis/batch/v1/defaults.go
- delete non-Job defaulting funcs

pkg/apis/batch/v1/defaults_test.go
- remove tests for non-Job.

pkg/apis/batch/v1/register.go
- remove non-job.

pkg/apis/batch/v1/types.go
- remove non-job, except define own copy of LabelSelector.

gofmt.

erictune committed:
Change extensions to batch and v1beta1 to v1.

Just in the newly added files.

erictune committed:
Added batch/v1 to code generators.

erictune committed:
Added installer for batch, copied from extensions.

erictune commented:
@kubernetes/goog-csi help please.

erictune committed:
Notes on stuff to do next.

erictune committed:
Register internal repr.

Steps:
  cp pkg/apis/extensions/register.go
  Delete things other than Job.
  Change Job to extensions.Job
  Change group name to Job
  Remove unused import.
  Add import.
  Delete already defined methods.

caesarxuchao commented:
I would say create the new `pkg/apis/batch/install/install.go`. 

To correctly generate the conversions, I think you need to register the internal types correctly, probably you should register the internal types again for the new group, with `SchemeGroupVersion` equal to your new group's internal version: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/register.go#L49.

lavalamp commented:
Yeah, that's what I suggested (IRL), too.

On Wed, Feb 3, 2016 at 2:48 PM, Chao Xu <notifications@github.com> wrote:

> I would say create the new pkg/apis/batch/install/install.go.
>
> To correctly generate the conversions, I think you need to register the
> internal types correctly, probably you should register the internal types
> again for the new group, with SchemeGroupVersion equal to your new
> group's internal version:
> https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/register.go#L49
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20551#issuecomment-179512585>
> .
>

erictune committed:
Edit install.go and install_test.go

Change tests from DaemonSet or HPA to Job.
Change extensions/v1beta1 to batch/v1.

erictune committed:
Add v1 package name to Convert_* functions.

These Convert*Pod* functions are copied from
legacy api because Pod conversions cannot be
automatically generated because of something
about deprecatedServiceAccount.

The way they were copied, they refer to other
conversion functions that aren't in the current
package except by accident of how genconversion
works.

Using the original package rather than local
allows running genconversion from a clean start.

lavalamp commented:
@piosz You may want to refer to this PR since you are doing the same thing just for different objects & groups.

erictune committed:
Generated conversion and deep copies.

hack/update-generated-conversion.sh
hack/update-generated-deep-copies.sh

erictune commented:
@googlebot you are confused.

erictune committed:
Fix to defaults_test.go

erictune committed:
Fix to pkg/apis/batch/install/install_test.go

erictune committed:
Handle batch/v1 in runtime-config

erictune committed:
Enable batch/v1 in tests.

erictune committed:
Fix import in pkg/apis/batch/v1/defaults_test.go.

erictune committed:
Add batch to cmd/libs/go2idl/client-gen/main.go

erictune commented:
@lavalamp @nikhiljindal 
I've been banging on this for hours but I am stuck.  

I cannot get `make test WHAT=pkg/apis/batch/v1` to pass.  The `defaults_test.go` is failing.  It looks like the roundTrip is returning an internal rather than a versioned type for the label selector.

Related to that, I cannot get genconversions to generate the functions `Convert_unversioned_LabelSelector_To_v1_LabelSelector` and `Convert_v1_LabelSelector_To_unversioned_LabelSelector`.  I worked around this by writing them by hand, but maybe this is the source of my first problem.

Any help appreciated.

BTW, I considered just copying what @piosz did in #20501 but HPA does not have selectors, so I need to sort out the selector conversion issue in any case.

lavalamp commented:
@erictune give me a few minutes to review #20501 to unblock @piosz and then let's chat

caesarxuchao commented:
Do you have LabelSelector defined in your versioned types.go? If you have, the conversion should be automatically generated.

erictune commented:
@caesarxuchao it is defined in the versioned.

caesarxuchao commented:
The test failure is because you miss the fuzzer:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testing/fuzzer.go#L159

erictune committed:
Generated a new pkg/apis/batch/v1/types.generated.go

erictune committed:
Updated pkg/master/master.go

erictune committed:
Added validation for new api group.

erictune committed:
Install batch/v1 to master and unversioned client.

erictune committed:
Take a stab at setting storage version.

But make test WHAT=pkg/master
fails

soltysh commented on pkg/apis/batch/v1/types.go:
```diff
@@ -0,0 +1,170 @@
+// A label selector is a label query over a set of resources. The result of matchLabels and
+// matchExpressions are ANDed. An empty label selector matches all objects. A null
+// label selector matches no objects.
+type LabelSelector struct {
```
Although `LabelSelector` is currently used only in Jobs, I'd rather have it in some general pkg than in `batch/`. Hopefully all the objects will use it in the future.

erictune replied:
@bgrant0607 @lavalamp Do we have to decide before 1.2 what the package is for the new LabelSelector struct that Job uses and ReplicaSet will use?  I realize it does not matter for our JSON api.  But, does it matter for pkg/client/typed/... that we have a stable package name for any types the client interacts with?  Or would we be allowed to change that with each new version of pkg/client/typed/.../version-blah/?

bgrant0607 replied:
Go packages and type names aren't part of the API. They are implementation details that are subject to change. That is also true for our "internal", unversioned client library, and any libraries in our main repo.

@lavalamp Will the generated versioned client be done in 1.2?

LabelSelector is currently in unversioned.

caesarxuchao replied:
> Will the generated versioned client be done in 1.2?

They are already in.
The versioned clientset is here: https://github.com/kubernetes/kubernetes/tree/master/pkg/client/clientset_generated/release_1_2
The versioned typed clients are at https://github.com/kubernetes/kubernetes/tree/master/pkg/client/typed/generated/core/v1 and https://github.com/kubernetes/kubernetes/tree/master/pkg/client/typed/generated/extensions/v1beta1

> does it matter for pkg/client/typed/... that we have a stable package name for any types the client interacts with?

The clientset doesn't interact with this LabelSelector, because it's not a top level API, so we can move LabelSelector later.

That said, we do require the types the generated client interacts with to stay in the same package after its release.

lavalamp replied:
Yes-- there's already a generated release_1_2 client. Take a look at the
doc Chao is adding: #21436

On Wed, Feb 17, 2016 at 8:26 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/apis/batch/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/20551#discussion_r53270274>
> :
>
> > +	// Status of the condition, one of True, False, Unknown.
> > +	Status v1.ConditionStatus `json:"status"`
> > +	// Last time the condition was checked.
> > +	LastProbeTime unversioned.Time `json:"lastProbeTime,omitempty"`
> > +	// Last time the condition transit from one status to another.
> > +	LastTransitionTime unversioned.Time `json:"lastTransitionTime,omitempty"`
> > +	// (brief) reason for the condition's last transition.
> > +	Reason string `json:"reason,omitempty"`
> > +	// Human readable message indicating details about last transition.
> > +	Message string `json:"message,omitempty"`
> > +}
> > +
> > +// A label selector is a label query over a set of resources. The result of matchLabels and
> > +// matchExpressions are ANDed. An empty label selector matches all objects. A null
> > +// label selector matches no objects.
> > +type LabelSelector struct {
>
> Go packages and type names aren't part of the API. They are implementation
> details that are subject to change. That is also true for our "internal",
> unversioned client library, and any libraries in our main repo.
>
> @lavalamp <https://github.com/lavalamp> Will the generated versioned
> client be done in 1.2?
>
> LabelSelector is currently in unversioned.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20551/files#r53270274>.
>

lavalamp replied:
Yeah, it's not perfect yet but it's getting much better-- the generated 1.2
client set at least uses the versioned types, not the internal types.

On Wed, Feb 17, 2016 at 9:45 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/apis/batch/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/20551#discussion_r53273947>
> :
>
> > +	// Status of the condition, one of True, False, Unknown.
> > +	Status v1.ConditionStatus `json:"status"`
> > +	// Last time the condition was checked.
> > +	LastProbeTime unversioned.Time `json:"lastProbeTime,omitempty"`
> > +	// Last time the condition transit from one status to another.
> > +	LastTransitionTime unversioned.Time `json:"lastTransitionTime,omitempty"`
> > +	// (brief) reason for the condition's last transition.
> > +	Reason string `json:"reason,omitempty"`
> > +	// Human readable message indicating details about last transition.
> > +	Message string `json:"message,omitempty"`
> > +}
> > +
> > +// A label selector is a label query over a set of resources. The result of matchLabels and
> > +// matchExpressions are ANDed. An empty label selector matches all objects. A null
> > +// label selector matches no objects.
> > +type LabelSelector struct {
>
> Will the generated versioned client be done in 1.2?
>
> They are already in.
> The versioned clientset is here:
> https://github.com/kubernetes/kubernetes/tree/master/pkg/client/clientset_generated/release_1_2
> The versioned typed clients are at
> https://github.com/kubernetes/kubernetes/tree/master/pkg/client/typed/generated/core/v1
> and
> https://github.com/kubernetes/kubernetes/tree/master/pkg/client/typed/generated/extensions/v1beta1
>
> does it matter for pkg/client/typed/... that we have a stable package name
> for any types the client interacts with?
>
> The clientset doesn't interact with this LabelSelector, because it's not a
> top level API, so we can move LabelSelector later.
>
> That said, we do require the types the generated client interacts with to
> stay in the same package after its release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20551/files#r53273947>.
>

soltysh commented on pkg/master/master.go:
```diff
@@ -253,6 +255,47 @@ func (m *Master) InstallAPIs(c *Config) {
 		allGroups = append(allGroups, group)
 	}
+	// Install batch unless disabled.
+	if !m.ApiGroupVersionOverrides["batch/v1"].Disable {
```
How about using job's v1.SchemeGroupVersion.String() instead of above string to remove repeating the strings?

erictune committed:
WIP.

erictune commented:
I'm removing the 1.2 milestone from this PR.  The feature is still targeted to 1.2 but the changes will be in a fresh PR.

erictune commented:
closing in favor of #21434
